Implementation of conditional dialogue and action-dialogue integrations; Wrote dialogue unit tests

jeyeong opened the pull request:
Introduced the following changes to the dialogue module.

**Conditional Dialogue**
- Allow dialogue designers to designate certain edges (i.e. dialogue options) as "conditional," that is, only accessible if the player has fulfilled certain conditions.
- The implementation makes use of conditions from game-state/conditions.h.
- To make start_conversation() and run_conversation_step() work with conditional dialogue, some revisions had to be made: (1) Implemented a function to check for which edges are "available" upon reaching the node. (2) Generating the option numbers in the return string based on these "available" nodes.
- The edge and edge_list struct also had to be revised.
- Multiple conditions are possible.

**Action Integrations**
- Allow nodes to have associated actions, for example, giving the player an item, or starting a quest.
- This was achieved by implementing a node_action struct (a linked list) and an associated do_node_actions() function. Upon reaching each node, do_node_actions() is called, executing each action.
- Multiple actions per node are possible.

**Unit Tests** (in npc/test_dialogue.c)
- Tests for the basic functions, such as the init, new, free functions, as well as add_node() and add_edge().
- Tests for dialogue execution, i.e. start_conversation() and run_conversation_step().
- Tests for conditional dialogue.
- Tests for actions.

A working example of conditional dialogue + actions can be found in **src/npc/examples/dialogue_example.c.**

pkasner commented on src/npc/src/dialogue.c:
```diff
@@ -185,35 +261,127 @@ char *create_return_string(convo_t *c, int is_leaf)
 
+    // Step 1: Traverse to the player's selected node
     cur_edge = c->cur_node->edges;
-    for (i = 1; i < input && i < c->cur_node->num_edges; i++) {
```
The logic in this step is a little difficult to follow. Not sure if that's something to clarify with more comments or just me being slow to process how the while conditions interact with the if statement. Perhaps specify that the input value associated with an edge is equal to its order in the linked list ignoring any unavailable edges before it.

jeyeong replied:
How did that get there? Was this from an old commit? The original idea of "i < c->cur_node->num_edges" was to handle the event where input is > than the number of available options. But I've since changed that implementation!

pkasner replied:
I think I clicked the wrong line with this comment. Check the code immediately underneath maybe?

jeyeong replied:
Ah, I see what you mean by hard to follow. This was the logic I needed to implement to make sure we end up at the correct edge. Perhaps I should state that explicitly in a comment.

pkasner commented on tests/npc/test_dialogue.c:
```diff
@@ -302,3 +307,368 @@ Test(dialogue, add_bidirectional_edges)
+}
+
+/* 2 normal edges and 2 conditional edges, one satisfied and one unsatisfied,
+   results in 2 available edges */
```
Some of these tests are probably redundant/don't seem to add additional coverage. No reason to delete them, but something to be aware of for the future.

jeyeong replied:
Thanks for pointing that out. My thought process for the conditional dialogue tests were as follows:
1. Conditional edge, not passed
2. Conditional edge, passed
3. 1 Unconditional edge + 1 Conditional edge, passed (testing for interaction between conditional, unconditional edges)
4. 2 Unconditional edges + 2 Conditional edges, 1 pass, 1 fail (testing for interaction between the two conditional edges)

Are you saying including test #4 alone would perhaps have sufficed?

pkasner replied:
I think 1 and 4 probably would have sufficed. 1 still seems like a useful edge case (as it returns 1 instead of 0 like the others).

jeyeong replied:
Got it, that makes sense. I've removed the redundant 2 & 3.

pkasner commented on include/npc/dialogue.h:
```diff
@@ -127,29 +168,67 @@ int add_edge(convo_t *c, char *quip, char *from_id, char *to_id);
+ *  - SUCCESS on success, FAILURE if an error occurs
+ *  - Possible errors: (1) node matching node_id could not be found;
+ *    (2) the node already has an action;
+ */
```
Because the actual initiating of quests is left as to do, so this should be fine for now.

jeyeong replied:
Ah, I see. Technically, it isn't completely out of the question to convert a string into a long int, and vice versa.

pkasner reviewed (changes requested):
For the most part looks good. Obviously the tests need to pass, so those will need to be figured out. Most important change is probably making sure item possession is handled gracefully for the npc and player if there is a transfer during dialogue. Also this perhaps would have been better as two pull requests, one for conditional dialogue and one for actions.

jeyeong commented on include/npc/dialogue.h:
```diff
@@ -12,6 +15,33 @@
+ */
+typedef struct node_action {
+    node_action_type action_type;
+    char *action_id;
```
Yes, that's correct. Action_type identifies the action being taken. Though it seems like that isn't so clear.

jeyeong replied:
I renamed action_type to simply action

jeyeong commented on src/npc/src/dialogue.c:
```diff
@@ -121,31 +115,111 @@ int add_edge(convo_t *c, char *quip, char *from_id, char *to_id)
+                item_t *item = get_item_in_hash(game->all_items,
+                                                action->action_id);
+                if (item == NULL) return FAILURE;
+                int rc = add_item_to_player(game->curr_player, item);
```
That's actually a good point. One thing we would have to provide if we want to remove the item from the NPC's inventory is a pointer to the NPC we are currently speaking to. Which we may have to think about.

jeyeong replied:
This current code itself, however, fails generally due to the item bug. I was considering fixing that error before moving on to a more proper implementation.

pkasner replied:
With regard to keeping track of the npc we're talking to, I notice this function takes the game struct as a parameter, which means it can access the game's mode. In conversation mode, the mode structs mode_ctx field contains the npc's npc_id.

jeyeong replied:
Ah, that would be convenient. Can you point me to the function that returns the npc struct using the npc_id?

pkasner replied:
There currently is no such function, but it can be done with the HASH_FIND uthash macro. However, this also assumes that the npc/mode_operation branch is merged, which adds the all_npcs field as a hast table to chiventure's game struct. I do think I will probably add a get_npc function or something of the sort to that branch in the npc module.

pkasner replied:
The get_npc() function in game.h in the npc/mode_operation branch now does that job.

jeyeong replied:
Great!

pkasner commented:
This looks good! I'm tentatively approving this, but I think we should still wait to merge this until after the npc/mode_operation branch can be merged. That way we can make sure conversation mode still works and we can do some integration testing.

pkasner reviewed (approved):
My above see my above comment that should've been attached to this review.

jeyeong commented:
After @pkasner 's pull request got merged, the code no longer compiled. I've made the changes so that it compiles now.

jeyeong commented:
I remodelled the GIVE_ITEM and TAKE_ITEM actions so that they interact with NPCs rather than the game. Also, since npc_new() was redefined in a recent update, I updated the example files so that they now work.

lilyehsani commented on src/npc/src/dialogue.c:
```diff
@@ -121,58 +115,157 @@ int add_edge(convo_t *c, char *quip, char *from_id, char *to_id)
-    totlen += (len_lst[i++] = strlen(c->cur_node->npc_dialogue));
-    totlen += 2;
+    totlen += (len_lst[i++] = strlen(n->npc_dialogue));
+    totlen += 1;
```
Was the change from 2 to 1 here intentional?

jeyeong replied:
It was, the reason is because previously, I had wanted two newlines to be printed. But I realized that the CLI doesn't handle multiple newlines and only prints 1, so I changed it to only have 1 newline.

lilyehsani replied:
Sounds good

lilyehsani commented on src/npc/src/dialogue.c:
```diff
@@ -185,35 +278,143 @@ char *create_return_string(convo_t *c, int is_leaf)
+
+    return add_action_to_node(n, START_QUEST, quest_id);
 }
 
```
Again, should probably write a to-do for add_start_battle

jeyeong replied:
Done

lilyehsani commented on src/npc/src/dialogue.c:
```diff
@@ -270,39 +469,27 @@ int edge_free(edge_t *e)
 node_t *node_new(char *node_id, char *npc_dialogue)
 {
-    assert(strlen(node_id) < MAX_NODE_ID_LEN);
-    assert(strlen(npc_dialogue) < MAX_DIA_LEN);
```
Why delete these?

jeyeong replied:
I realized that I was doing assertion testing twice: once in new() and once in init(). So I removed the assertions in new() and kept them in init()

lilyehsani replied:
Ah I see

lilyehsani commented on src/npc/src/dialogue.c:
```diff
@@ -389,3 +579,33 @@ int free_node_list(node_list_t *n_lst, bool free_nodes)
 }
+
+/* See dialogue.h */
+node_action_t *node_action_new(node_action_type action, char *action_id)
```
Should write corresponding init/free functions for this.

jeyeong replied:
Done

jeyeong replied:
The free function was free_action_list() but I guess that wasn't very clear. I've renamed it to free_node_actions()

lilyehsani reviewed (changes requested):
Really great job overall! Added a few comments, many of which are just questions.

jeyeong commented on include/npc/dialogue.h:
```diff
@@ -22,39 +53,45 @@ typedef struct node node_t;
  *  - next, prev: next and previous list elements
  */
 typedef struct edge_list {
-    int option_number;
```
No, I've updated the create_return_string() function so it works independent of `option_number`.

lilyehsani replied:
Gotcha

jeyeong commented on include/npc/dialogue.h:
```diff
@@ -127,29 +167,78 @@ int add_edge(convo_t *c, char *quip, char *from_id, char *to_id);
+ *  - SUCCESS on success, FAILURE if an error occurs
+ *  - Possible errors: (1) node matching node_id could not be found;
+ */
+int add_start_quest(convo_t *c, char *node_id, char *quest_id);
```
I see. Would it be better to: (1) convert to long int into a string (using sprintf), or (2) create a union struct that could hold different types of IDs?

jeyeong replied:
Actually, I'll just leave it as a to-do for now, since I don't believe the quest team has fully implemented quests into Chiventure

jeyeong replied:
Actually, I'll just leave it as a to-do for now, since I don't believe we are able to integrate quests with dialogue at this point in time

lilyehsani replied:
Sounds good

jeyeong commented on include/npc/dialogue.h:
```diff
@@ -127,29 +167,78 @@ int add_edge(convo_t *c, char *quip, char *from_id, char *to_id);
+ */
+int add_start_quest(convo_t *c, char *node_id, char *quest_id);
 
 
```
Done

jeyeong commented on src/npc/examples/npc_example.c:
```diff
@@ -30,7 +30,7 @@ convo_t *create_sample_convo()
     // Starting to build the conversation structure
     convo_t *c = convo_new();
```
Done

jeyeong replied:
I only removed the commented code, instead of deleting the file because I think this was one of the files Prof. Sotomayor wanted students to refer to to get an idea of how monkeypatching works.

lilyehsani replied:
Awesome!

lilyehsani reviewed (approved):
LGTM! Approved.

jeyeong commented:
Summary of changes made:
- Added conditional dialogue functionality to dialogues.
- Added actions to dialogue (e.g. you can gain an item from a dialogue node).
- Wrote dialogue unit tests.

Next steps:
- Complete the load_npc module, so that NPCs can be loaded from WDL files.
- Write a working game demo featuring NPCs and dialogue.

lilyehsani commented:
Issue Score: ✔️+

Comments:
Great work overall! I have one comment:
- This PR represents a large amount of work that could've been integrated into `dev` through smaller PRs earlier in the quarter. Remember that you should not treat PRs as a singular event to merge all or most of your work into `dev`. Integrating incrementally can help identify integration issues early, as well as ensure that you are not rushing to merge all your code towards the end of a sprint. For example, the conditional dialogue, dialogue actions, and unit tests could have been separated into 3 or some combination of 2 PRs.

<!-- {"score": "check-plus"} -->
